src/router/privoxy/privoxy.1
.\" This manpage has been generated by docbook2man-spec.pl
.\" (included in the Privoxy source tarball) from a DocBook document.
.TH "PRIVOXY" "1" "26 August 2016" "Privoxy 3.0.26" ""
.SH NAME
privoxy \- Privacy Enhancing Proxy
.SH SYNOPSIS

\fBprivoxy\fR [\fB\-\-chroot\fR ] [\fB\-\-config-test\fR ] [\fB\-\-help\fR ] [\fB\-\-no-daemon\fR ] [\fB\-\-pidfile \fIpidfile\fB\fR ] [\fB\-\-pre-chroot-nslookup \fIhostname\fB\fR ] [\fB\-\-user \fIuser[.group]\fB\fR ] [\fB\-\-version\fR ] [\fB\fIconfigfile\fB\fR ]

.SH "OPTIONS"
.PP
\fBPrivoxy\fR may be invoked with the following command line
options:
.TP
\fB\-\-chroot\fR
Before changing to the user ID given in the \-\-user option, chroot to
that user's home directory, i.e. make the kernel pretend to the
\fBPrivoxy\fR process that the directory tree starts
there. If set up carefully, this can limit the impact of possible
vulnerabilities in \fBPrivoxy\fR to the files contained in
that hierarchy.
.TP
\fB\-\-config-test\fR
Exit after loading the configuration files before binding to
the listen address. The exit code signals whether or not the
configuration files have been successfully loaded.

If the exit code is 1, at least one of the configuration files
is invalid, if it is 0, all the configuration files have been
successfully loaded (but may still contain errors that can
currently only be detected at run time).

This option doesn't affect the log setting, combination with
"--no-daemon" is recommended if a configured log file shouldn't
be used.
.TP
\fB\-\-help\fR
Print brief usage info and exit.
.TP
\fB\-\-no-daemon\fR
Don't  become  a daemon, i.e. don't fork and become process group
leader, don't detach from controlling tty, and do all logging there.
.TP
\fB\-\-pidfile \fIpidfile\fB\fR
On startup, write the process ID to \fIpidfile\fR.
Delete the \fIpidfile\fR on exit.
Failure to create or delete the \fIpidfile\fR
is non-fatal. If no \fB\-\-pidfile\fR option is given, no PID file will be used.
.TP
\fB\-\-pre-chroot-nslookup \fIhostname\fB\fR
Initialize the resolver library using \fIhostname\fR
before chroot'ing. On some systems this reduces the number of files
that must be copied into the chroot tree.
.TP
\fB\-\-user \fIuser[.group]\fB\fR
After (optionally) writing the PID file, assume the user ID of
\fIuser\fR and the GID of
\fIgroup\fR, or, if the optional
\fIgroup\fR was not given, the default group of
\fIuser\fR. Exit if the privileges are not
sufficient to do so.
.TP
\fB\-\-version\fR
Print version info and exit.
.PP
If the \fIconfigfile\fR is not specified on  the  command  line,
\fBPrivoxy\fR  will  look for a file named
\fIconfig\fR in the current directory. If no
\fIconfigfile\fR is found, \fBPrivoxy\fR will
fail to start.
.SH "DESCRIPTION"
.PP
Privoxy is a non-caching web proxy with advanced filtering capabilities
for enhancing privacy, modifying web page data and HTTP headers, controlling
access, and removing ads and other obnoxious Internet junk. Privoxy has a
flexible configuration and can be customized to suit individual needs and tastes.
It has application for both stand-alone systems and multi-user networks.
.PP
Privoxy is Free Software and licensed under the GNU GPLv2.
.PP
Privoxy is an associated project of Software in the Public Interest (SPI).
.PP
Helping hands and donations are welcome:
.TP 0.2i
\(bu
https://www.privoxy.org/faq/general.html#PARTICIPATE
.TP 0.2i
\(bu
https://www.privoxy.org/faq/general.html#DONATE
.SH "INSTALLATION AND USAGE"
.PP
Browsers can either be individually configured to use
\fBPrivoxy\fR as a HTTP proxy (recommended),
or \fBPrivoxy\fR can be combined with a packet
filter to build an intercepting proxy
(see \fIconfig\fR).  The default setting is  for
localhost,  on port  8118 (configurable in the main config file).  To set the
HTTP proxy in Firefox, go through: \fBTools\fR;
\fBOptions\fR; \fBGeneral\fR;
\fBConnection Settings\fR;
\fBManual Proxy Configuration\fR.
.PP
For Internet Explorer, go through: \fBTools\fR;
\fBInternet Properties\fR; \fBConnections\fR;
\fBLAN Settings\fR.
.PP
The Secure (SSL) Proxy should also be set to the same values, otherwise
https: URLs will not be proxied. Note: \fBPrivoxy\fR can only
proxy HTTP and HTTPS traffic. Do not try it with FTP or other protocols.
HTTPS presents some limitations, and not all features will work with HTTPS
connections.
.PP
For other browsers, check the documentation.
.SH "CONFIGURATION"
.PP
\fBPrivoxy\fR can be configured with the various configuration
files. The default configuration files are: \fIconfig\fR,
\fIdefault.filter\fR, \fIdefault.action\fR and
\fIdefault.action\fR. \fIuser.action\fR should
be used for locally defined exceptions to the default rules in
\fImatch-all.action\fR and \fIdefault.action\fR,
and \fIuser.filter\fR for locally defined filters. These are
well commented.  On Unix and Unix-like systems, these are located in
\fI/etc/privoxy/\fR by default.
.PP
\fBPrivoxy\fR uses the concept of \fBactions\fR
in order to manipulate the data stream between the browser and remote sites.
There are various actions available with specific functions for such things
as blocking web sites, managing cookies, etc. These actions can be invoked
individually or combined, and used against individual URLs, or groups of URLs
that can be defined using wildcards and regular expressions. The result is
that the user has greatly enhanced control and freedom.
.PP
The actions list (ad blocks, etc) can also be configured with your
web browser at http://config.privoxy.org/
(assuming the configuration allows it).
\fBPrivoxy's\fR configuration parameters  can also  be viewed at
the same page. In addition, \fBPrivoxy\fR can be toggled on/off.
This is an internal page, and does not require Internet access.
.PP
See the \fIUser Manual\fR for a detailed
explanation of installation, general usage, all configuration options, new
features and notes on upgrading.
.SH "FILES"

.nf
 \fI/usr/sbin/privoxy\fR
 \fI/etc/privoxy/config\fR
 \fI/etc/privoxy/match-all.action\fR
 \fI/etc/privoxy/default.action\fR
 \fI/etc/privoxy/user.action\fR
 \fI/etc/privoxy/default.filter\fR
 \fI/etc/privoxy/user.filter\fR
 \fI/etc/privoxy/trust\fR
 \fI/etc/privoxy/templates/*\fR
 \fI/var/log/privoxy/logfile\fR
.fi
.PP
Various other files should be included, but may vary depending on platform
and build configuration. Additional documentation should be included in the local
documentation directory.
.SH "SIGNALS"
.PP
\fBPrivoxy\fR terminates on the \fBSIGINT\fR
and \fBSIGTERM\fR signals. Log
rotation scripts may cause a re-opening of the logfile by sending a
\fBSIGHUP\fR to \fBPrivoxy\fR. Note that unlike
other daemons,  \fBPrivoxy\fR does not need to be made aware of
config file changes by \fBSIGHUP\fR -- it will detect them
automatically. Signals other than the ones listed above aren't explicitly
handled and result in the default action defined by the operating system.
.SH "NOTES"
.PP
Please see the \fIUser Manual\fR on how to contact the
developers, for feature requests, reporting problems, and other questions.
.SH "SEE ALSO"
.PP
Other references and sites of interest to \fBPrivoxy\fR
users:
.PP

https://www.privoxy.org/,
the \fBPrivoxy\fR Home page.

https://www.privoxy.org/faq/,
the \fBPrivoxy\fR FAQ.

https://www.privoxy.org/developer-manual/,
the \fBPrivoxy\fR developer manual.

https://sourceforge.net/projects/ijbswa/,
the Project Page for \fBPrivoxy\fR on
SourceForge.

http://config.privoxy.org/,
the web-based user interface. \fBPrivoxy\fR must be
running for this to work. Shortcut: http://p.p/

https://sourceforge.net/tracker/?group_id=11118&atid=460288, to submit ``misses'' and other
configuration related suggestions to the developers.
.SH "DEVELOPMENT TEAM"

.nf
 Fabian Keil, lead developer
 David Schmidt
 Hal Burgiss
 Lee Rian
 Roland Rosenfeld
 Ian Silvester
.fi
.SH "COPYRIGHT AND LICENSE"
.SS "COPYRIGHT"
.PP
Copyright (C) 2001-2016 by Privoxy Developers <privoxy-devel@lists.privoxy.org>
.PP
Some source code is based on code Copyright (C) 1997 by Anonymous Coders
and Junkbusters, Inc. and licensed under the \fIGNU General Public
License\fR.
.SS "LICENSE"
.PP
\fBPrivoxy\fR is free software; you can
redistribute it and/or modify it under the terms of the
\fIGNU General Public License\fR, version 2,
as published by the Free Software Foundation.
.PP
\fBPrivoxy\fR is distributed in the hope that
it will be useful, but WITHOUT ANY WARRANTY; without even the implied
warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
See the  \fIlicense\fR for details.


src/router/services/services/privoxy.c

#ifdef HAVE_PRIVOXY
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <errno.h>
#include <sys/time.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <syslog.h>
#include <signal.h>
#include <utils.h>
#include <bcmnvram.h>
#include <shutils.h>
#include <services.h>

void start_privoxy(void)
{
	if (!nvram_matchi("privoxy_enable", 1))
		return;

	int mode = 0;
	int whitelist = 0;
	char *next;
	char var[80];
	char vifs[256];
	char *ip = nvram_safe_get("lan_ipaddr");
	char *mask = nvram_safe_get("lan_netmask");
	char *webif_port = nvram_safe_get("http_lanport");

	sysprintf("grep -q nobody /etc/passwd || echo \"nobody:*:65534:65534:nobody:/var:/bin/false\" >> /etc/passwd");
	mkdir("/var/log/privoxy", 0777);

	char *wan = get_wan_ipaddr();
	if (nvram_matchi("privoxy_transp_enable", 1)) {
		sysprintf("iptables -t nat -D PREROUTING -p tcp -d ! %s --dport 80 -j DNAT --to %s:8118", wan, ip);
		sysprintf("iptables -t nat -I PREROUTING -p tcp -d ! %s --dport 80 -j DNAT --to %s:8118", wan, ip);
		sysprintf("iptables -t nat -D PREROUTING -p tcp -s %s/%s -d %s --dport %s -j ACCEPT", ip, mask, ip, webif_port);
		sysprintf("iptables -t nat -I PREROUTING -p tcp -s %s/%s -d %s --dport %s -j ACCEPT", ip, mask, ip, webif_port);
		sysprintf("iptables -t nat -D PREROUTING -p tcp -s %s -d %s --dport %s -j DROP", ip, ip, webif_port);
		sysprintf("iptables -t nat -I PREROUTING -p tcp -s %s -d %s --dport %s -j DROP", ip, ip, webif_port);
		/* no gui setting yet - redirect all except this IP */
		if (strlen(nvram_safe_get("privoxy_transp_exclude"))) {
			sysprintf("iptables -t nat -D PREROUTING -p tcp -s %s --dport 80 -j ACCEPT", nvram_safe_get("privoxy_transp_exclude"));
			sysprintf("iptables -t nat -I PREROUTING -p tcp -s %s --dport 80 -j ACCEPT", nvram_safe_get("privoxy_transp_exclude"));
		}
		mode = 1;
		getIfLists(vifs, 256);
		char vif_ip[32];
		foreach(var, vifs, next) {
			if (strcmp(get_wan_face(), var)
			    && strcmp(nvram_safe_get("lan_ifname"), var)) {
				if (nvram_nmatch("1", "%s_isolation", var)) {
					sysprintf("iptables -t nat -D PREROUTING -i %s -d %s/%s -j RETURN", var, ip, mask);
					sysprintf("iptables -t nat -I PREROUTING -i %s -d %s/%s -j RETURN", var, ip, mask);
					sprintf(vif_ip, "%s_ipaddr", var);
					sysprintf("iptables -t nat -D PREROUTING -i %s -d %s -j RETURN", var, nvram_safe_get(vif_ip));
					sysprintf("iptables -t nat -I PREROUTING -i %s -d %s -j RETURN", var, nvram_safe_get(vif_ip));
				}
			}
		}

	}

	if (nvram_invmatch("privoxy_whitel", "")) {
		eval("cp", "/etc/privoxy/user.action", "/tmp/user.action");
		FILE *fpuser = fopen("/tmp/user.action", "ab");
		fprintf(fpuser, "%s", "{ -block }\n");
		fprintf(fpuser, "%s", nvram_safe_get("privoxy_whitel"));
		fclose(fpuser);
		whitelist = 1;
	}

	FILE *fp = fopen("/tmp/privoxy.conf", "wb");

	if (nvram_matchi("privoxy_advanced", 1) && nvram_invmatch("privoxy_conf", "")) {
		fprintf(fp, "%s", nvram_safe_get("privoxy_conf"));
	} else {
		fprintf(fp, "confdir /etc/privoxy\n"
			"logdir /var/log/privoxy\n"
			"actionsfile match-all.action\n"
			"actionsfile default.action\n"
			"actionsfile %s\n"
			"filterfile default.filter\n"
			"filterfile user.filter\n"
			"logfile logfile\n"
			"listen-address  %s:8118\n"
			"toggle  1\n"
			"enable-remote-toggle  0\n"
			"enable-remote-http-toggle  0\n"
			"enable-edit-actions 0\n"
			"buffer-limit 4096\n"
			"accept-intercepted-requests %d\n"
			"split-large-forms 0\n" "socket-timeout 60\n" "handle-as-empty-doc-returns-ok 1\n", whitelist ? "/tmp/user.action" : "user.action", ip, mode);
	}
	fclose(fp);
	eval("privoxy", "/tmp/privoxy.conf");
	syslog(LOG_INFO, "Privoxy : privoxy started\n");
	return;
}

void stop_privoxy(void)
{
	char *ip = nvram_safe_get("lan_ipaddr");
	char *wan = get_wan_ipaddr();
	char *mask = nvram_safe_get("lan_netmask");
	char *webif_port = nvram_safe_get("http_lanport");
	char *next;
	char var[80];
	char vifs[256];

	sysprintf("iptables -t nat -D PREROUTING -p tcp -d ! %s --dport 80 -j DNAT --to %s:8118", wan, ip);
	sysprintf("iptables -t nat -D PREROUTING -p tcp -s %s/%s -d %s --dport %s -j ACCEPT", ip, mask, ip, webif_port);
	sysprintf("iptables -t nat -D PREROUTING -p tcp -s %s -d %s --dport %s -j DROP", ip, ip, webif_port);
	sysprintf("iptables -t nat -D PREROUTING -p tcp -s %s --dport 80 -j ACCEPT", nvram_safe_get("privoxy_transp_exclude"));

	getIfLists(vifs, 256);
	char vif_ip[32];
	foreach(var, vifs, next) {
		if (strcmp(get_wan_face(), var)
		    && strcmp(nvram_safe_get("lan_ifname"), var)) {
			if (nvram_nmatch("1", "%s_isolation", var)) {
				sysprintf("iptables -t nat -D PREROUTING -i %s -d %s/%s -j RETURN", var, ip, mask);
				sprintf(vif_ip, "%s_ipaddr", var);
				sysprintf("iptables -t nat -D PREROUTING -i %s -d %s -j RETURN", var, nvram_safe_get(vif_ip));
			}
		}
	}
	stop_process("privoxy", "privoxy");
}
#endif
